// reakit/lib/Menu/StaticMenu.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var warning = require('reakit-utils/warning');
var React = require('react');
require('reakit-utils/useId');
var createComponent = require('reakit-system/createComponent');
var createHook = require('reakit-system/createHook');
var Box = require('../Box/Box.js');
var __chunk_1 = require('../chunk-002fc018.js');
require('reakit-utils/mergeRefs');
require('reakit-utils/useLiveRef');
var useAllCallbacks = require('reakit-utils/useAllCallbacks');
require('../Tabbable/Tabbable.js');
require('reakit-utils/useSealedState');
var useCreateElement = require('reakit-system/useCreateElement');
var usePipe = require('reakit-utils/usePipe');
require('../Hidden/HiddenState.js');
require('reakit-utils/useUpdateEffect');
require('../Dialog/DialogState.js');
require('reakit-utils/createOnKeyDown');
require('../Rover/RoverState.js');
require('../Rover/Rover.js');
require('popper.js');
require('../Popover/PopoverState.js');
var __chunk_7 = require('../chunk-8a7a2ea8.js');
var MenuState = require('./MenuState.js');

function useShortcuts(_ref, timeout) {
  var stops = _ref.stops,
      move = _ref.move;

  if (timeout === void 0) {
    timeout = 500;
  }

  var _React$useState = React.useState(""),
      keys = _React$useState[0],
      setKeys = _React$useState[1];

  React.useEffect(function () {
    if (!keys) return undefined;
    var timeoutId = setTimeout(function () {
      return setKeys("");
    }, timeout);
    var stop = stops.find(function (s) {
      return Boolean(s.ref.current && s.ref.current.textContent && s.ref.current.textContent.toLowerCase().startsWith(keys));
    });

    if (stop) {
      move(stop.id);
    }

    return function () {
      return clearTimeout(timeoutId);
    };
  }, [keys, stops, move]);
  var onKeyDown = React.useCallback(function (event) {
    if (event.metaKey || event.altKey || event.shiftKey || event.ctrlKey) {
      return;
    }

    if (/^[a-z0-9_-]$/i.test(event.key)) {
      event.stopPropagation();
      event.preventDefault();
      setKeys("" + keys + event.key);
    }
  }, [keys, setKeys]);
  return onKeyDown;
}

var useStaticMenu = createHook.createHook({
  name: "StaticMenu",
  compose: Box.useBox,
  useState: MenuState.useMenuState,
  useProps: function useProps(options, _ref) {
    var htmlOnKeyDown = _ref.onKeyDown,
        htmlWrap = _ref.unstable_wrap,
        htmlProps = __chunk_1._objectWithoutPropertiesLoose(_ref, ["onKeyDown", "unstable_wrap"]);

    var parent = React.useContext(__chunk_7.MenuContext);
    var providerValue = React.useMemo(function () {
      return {
        orientation: options.orientation,
        next: options.next,
        previous: options.previous,
        parent: parent
      };
    }, [options.orientation, options.next, options.previous, parent]);
    var onKeyDown = useShortcuts(options);
    var wrap = React.useCallback(function (children) {
      return React.createElement(__chunk_7.MenuContext.Provider, {
        value: providerValue
      }, children);
    }, [providerValue]);
    return __chunk_1._objectSpread({
      role: options.orientation === "horizontal" ? "menubar" : "menu",
      "aria-orientation": options.orientation,
      onKeyDown: useAllCallbacks.useAllCallbacks(onKeyDown, htmlOnKeyDown),
      unstable_wrap: usePipe.usePipe(wrap, htmlWrap)
    }, htmlProps);
  }
});
var StaticMenu = createComponent.createComponent({
  as: "div",
  useHook: useStaticMenu,
  useCreateElement: function useCreateElement$1(type, props, children) {
    process.env.NODE_ENV !== "production" ? warning.warning(!props["aria-label"] && !props["aria-labelledby"] && props.role !== "menubar", "Menu", "You should provide either `aria-label` or `aria-labelledby` props.", "See https://reakit.io/docs/menu") : void 0;
    return useCreateElement.useCreateElement(type, props, children);
  }
});

exports.StaticMenu = StaticMenu;
exports.useStaticMenu = useStaticMenu;
