// reakit/lib/Form/FormLabel.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('react');
require('reakit-utils/useId');
var createComponent = require('reakit-system/createComponent');
var createHook = require('reakit-system/createHook');
var Box = require('../Box/Box.js');
var __chunk_1 = require('../chunk-002fc018.js');
require('reakit-utils/useSealedState');
require('reakit-utils/useUpdateEffect');
require('reakit-utils/isPromise');
require('reakit-utils/isEmpty');
require('reakit-utils/isObject');
var FormState = require('./FormState.js');
require('./utils/setAllIn.js');
require('./utils/getIn.js');
require('reakit-utils/toArray');
require('reakit-utils/isInteger');
require('./utils/setIn.js');
require('../chunk-0ecca8fe.js');
var __chunk_3 = require('../chunk-79808ef5.js');
var __chunk_4 = require('../chunk-c9503aa1.js');

var unstable_useFormLabel = createHook.createHook({
  name: "FormLabel",
  compose: Box.useBox,
  useState: FormState.unstable_useFormState,
  keys: ["name", "label"],
  useProps: function useProps(options, htmlProps) {
    return __chunk_1._objectSpread({
      children: options.label,
      id: __chunk_4.getLabelId(options.name, options.baseId),
      htmlFor: __chunk_3.getInputId(options.name, options.baseId)
    }, htmlProps);
  }
});
var unstable_FormLabel = createComponent.createComponent({
  as: "label",
  useHook: unstable_useFormLabel
});

exports.unstable_FormLabel = unstable_FormLabel;
exports.unstable_useFormLabel = unstable_useFormLabel;
